### PR TITLE
feat: add safeguards for partially validated transactions [part 2/2]

### PR DESCRIPTION
**Pull request by jansegre**
This change was split out from #577. The main objective is to be able to do this:

```python
tx.set_validation(ValidationState.BASIC)  # this is a partially validated transaction
tx_storage.save_transaction(tx)  # this fails with an assertion error
with stx_storage.allow_partially_validated_context():
    tx_storage.save_transaction(tx)  # but this succeeds
tx_storage.get_transaction(tx.hash)  # this also fails, but with an error that inherits from "not found"
with stx_storage.allow_partially_validated_context():
    tx_storage.get_transaction(tx.hash)  # but this succeeds
```

This change is only exists to reduce the scope where the storage APIs accept reading/writing partially validated transactions so that the changes needed by sync-v2 can have a very reduced an limited impact on the rest of the node. The previous iteration of this protection was made using a `allow_partially_validated: bool` argument on most APIs, that ended up being not only verbose, but more error prone and needed to carry this `allow_partially_validated` in a lot of places that shouldn't have to know about that.

## Acceptance Criteria

- TransactionStorage must have a scope that dictates which validation level on accept at save/get methods
- TransactionStorage must have methods that return contextmanagers which temporarily modify that scope

### CODE REVIEW

File: hathor/transaction/storage/tx_allow_scope.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    purpose of filtering it can be any combination of the supported types.
+    """
+    VALID = auto()
+    PARTIAL = auto()
+    INVALID = auto()
+
+    def is_allowed(self, tx: BaseTransaction) -> bool:
+        """True means it is allowed to be used in the storage (as argument or as return), False means not allowed."""
+        tx_meta = tx.get_metadata()
+        is_partial = tx_meta.voided_by is not None and settings.PARTIALLY_VALIDATED_ID in tx_meta.voided_by
```
Comment (msbrogli) [inline hathor/transaction/storage/tx_allow_scope.py:42]: Why don't we use `tx_meta.validation`?

File: hathor/transaction/storage/transaction_storage.py [hunk 2]
```diff
@@ -330,6 +339,43 @@ def get_transaction_from_weakref(self, hash_bytes: bytes) -> Optional[BaseTransa
...
+        """This method is used to temporarily make the storage allow invalid transactions.
+
+        The implementation will INCLUDE allowing invalid transactions to the current allow scope.
+        """
+        new_allow_scope = self.allow_scope | TxAllowScope.INVALID
+        return tx_allow_context(self, allow_scope=new_allow_scope)
+
+    def is_only_valid_allowed(self) -> bool:
+        """Whether only valid transactions are allowed to be returned/accepted by the storage, the default state."""
+        return TxAllowScope.INVALID in self.allow_scope
```
Comment (msbrogli) [inline hathor/transaction/storage/transaction_storage.py:369]: Bug here.

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/transaction/storage/transaction_storage.py (hunk 2 above)
Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py]: Fixed.

File: hathor/transaction/storage/tx_allow_scope.py (hunk 1 above)
Comment (jansegre) [reply hathor/transaction/storage/tx_allow_scope.py]: Done.

File: hathor/transaction/storage/tx_allow_scope.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    VALID = auto()
+    PARTIAL = auto()
+    INVALID = auto()
+
+    def is_allowed(self, tx: BaseTransaction) -> bool:
+        """True means it is allowed to be used in the storage (as argument or as return), False means not allowed."""
+        tx_meta = tx.get_metadata()
+        # XXX: initial/partial/invalid/fully_connected never overlap and cover all possible validation states
+        #      see hathor.transaction.transaction_metadata.ValidationState for more details
+        validation = tx_meta.validation
```
Comment (msbrogli) [inline hathor/transaction/storage/tx_allow_scope.py:44]: Should we add `assert validation in list_with_all_possible_states`?

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/transaction/storage/transaction_storage.py (hunk 2 above)
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py]: Thanks!

File: hathor/transaction/storage/tx_allow_scope.py (hunk 1 above)
Comment (msbrogli) [reply hathor/transaction/storage/tx_allow_scope.py]: Thanks!

File: hathor/transaction/storage/transaction_storage.py
Comment (msbrogli) [inline hathor/transaction/storage/transaction_storage.py:1]: Should `get_metadata()` assert that the validation state and `meta.voided_by` are coherent?

Comment (msbrogli) [inline hathor/transaction/storage/transaction_storage.py:1]: Shouldn't `save_transaction(...)` follow the scope? For clarity, if one is trying to save an invalid transaction, this transaction must be allowed by the current scope.

File: hathor/transaction/storage/tx_allow_scope.py (hunk 3 above)
Comment (jansegre) [reply hathor/transaction/storage/tx_allow_scope.py:44]: In this case `list_with_all_possible_states = list(ValidationState)`. Asserting that in itself is pretty redundant with the whole Enum class.

What maybe wouldn't be redundant is `assert list(ValidationState) = [ValidationState.INITIAL, ValidationState.BASIC, ...]`, which would basically assert what the expected enum variants are. I'm not sure if it makes sense to do this here though. Maybe it would make more sense to have a simple test for that which could point to the known code locations that make certain assumptions based on the enum variants. I'm not too sure how much this helps though.

What do you think?

File: hathor/transaction/storage/transaction_storage.py
Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py:1]: I think that makes sense, I'll do that.

Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py:1]: Yes. I'll do that.

Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py:1]: That is already the case now since `post_get_validation` happens when `get_metadata` calls `get_transaction`.

Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py:1]: Done. Also refactored the tests to better cover this.

**Review by msbrogli — APPROVED** [review]

File: hathor/transaction/storage/transaction_storage.py
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py:1]: Great job! Thanks!

Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py:1]: Great job! Thanks!

File: hathor/transaction/storage/tx_allow_scope.py (hunk 3 above)
Comment (msbrogli) [reply hathor/transaction/storage/tx_allow_scope.py:44]: `assert list(ValidationState) == [ValidationState.INITIAL, ...]` is exactly what I meant. But I guess we can pass it. It would be great to have a mechanism that guarantees we have covered all possible cases otherwise a linter would fail or an exception would be raised.

Comment (jansegre) [reply hathor/transaction/storage/tx_allow_scope.py:44]: I'm actually tweaking this a little bit because of some edge cases (when using full validations we temporarily save transactions with INITIAL state). I'll add a test to assert that assumption (in the XXX comment) and the state list.

Comment (jansegre) [reply hathor/transaction/storage/tx_allow_scope.py:44]: Done. I've added those tests.

File: hathor/manager.py [hunk 4]
```diff
@@ -441,13 +441,15 @@ def _initialize_components(self) -> None:
         # self.start_profiler()
         if self._full_verification:
             self.log.debug('reset all metadata')
-            for tx in self.tx_storage.get_all_transactions():
-                tx.reset_metadata()
+            with self.tx_storage.allow_partially_validated_context():
+                for tx in self.tx_storage.get_all_transactions():
```
Comment (msbrogli) [inline hathor/manager.py:445]: Should `get_all_transactions()` only return transactions in the scope?

File: hathor/manager.py [hunk 5]
```diff
@@ -441,13 +441,15 @@ def _initialize_components(self) -> None:
...
+                    tx.reset_metadata()
 
         self.log.debug('load blocks and transactions')
         for tx in self.tx_storage._topological_sort_dfs():
```
Comment (msbrogli) [inline hathor/manager.py:449]: Same thing here: should it go through only transactions in the scope?

File: hathor/manager.py [hunk 6]
```diff
@@ -441,13 +441,15 @@ def _initialize_components(self) -> None:
...
             if self._full_verification:
-                tx.update_initial_metadata()
+                with self.tx_storage.allow_partially_validated_context():
+                    tx.update_initial_metadata()
```
Comment (msbrogli) [inline hathor/manager.py:452]: Why would this access need a broader scope?

File: hathor/manager.py [hunk 7]
```diff
@@ -480,16 +482,19 @@ def _initialize_components(self) -> None:
                     # TODO: deal with invalid tx
                     if tx.can_validate_full():
                         self.tx_storage.add_to_indexes(tx)
-                        assert tx.validate_full(skip_block_weight_verification=skip_block_weight_verification)
+                        assert tx.validate_full(skip_block_weight_verification=skip_block_weight_verification,
+                                                reject_locked_reward=False)
```
Comment (msbrogli) [inline hathor/manager.py:486]: Why are we adding `reject_locked_reward=False` here?

File: hathor/manager.py [hunk 8]
```diff
@@ -480,16 +482,19 @@ def _initialize_components(self) -> None:
...
                         self.consensus_algorithm.update(tx)
                         self.tx_storage.indexes.update(tx)
                         if self.tx_storage.indexes.mempool_tips is not None:
                             self.tx_storage.indexes.mempool_tips.update(tx)  # XXX: move to indexes.update
                         if self.tx_storage.indexes.deps is not None:
                             self.sync_v2_step_validations([tx])
+                        self.tx_storage.save_transaction(tx, only_metadata=True)
```
Comment (msbrogli) [inline hathor/manager.py:493]: Why save here?

File: hathor/manager.py [hunk 9]
```diff
@@ -480,16 +482,19 @@ def _initialize_components(self) -> None:
...
                     else:
-                        assert tx.validate_basic(skip_block_weight_verification=skip_block_weight_verification)
-                    self.tx_storage.save_transaction(tx, only_metadata=True)
+                        with self.tx_storage.allow_partially_validated_context():
+                            assert tx.validate_basic(skip_block_weight_verification=skip_block_weight_verification)
+                            self.tx_storage.save_transaction(tx, only_metadata=True)
```
Comment (msbrogli) [inline hathor/manager.py:497]: Why save here?

File: hathor/manager.py [hunk 10]
```diff
@@ -480,16 +482,19 @@ def _initialize_components(self) -> None:
...
                         self.tx_storage.indexes.update(tx)
                         if self.tx_storage.indexes.mempool_tips is not None:
                             self.tx_storage.indexes.mempool_tips.update(tx)  # XXX: move to indexes.update
                         if self.tx_storage.indexes.deps is not None:
                             self.sync_v2_step_validations([tx])
+                        self.tx_storage.save_transaction(tx, only_metadata=True)
                     else:
-                        assert tx.validate_basic(skip_block_weight_verification=skip_block_weight_verification)
-                    self.tx_storage.save_transaction(tx, only_metadata=True)
+                        with self.tx_storage.allow_partially_validated_context():
```
Comment (msbrogli) [inline hathor/manager.py:495]: Why do we need access to partially validated transactions to run a basic validation?

File: hathor/consensus/consensus.py [hunk 11]
```diff
@@ -111,7 +113,8 @@ def _unsafe_update(self, base: BaseTransaction) -> None:
             if to_remove:
                 self.log.warn('some transactions on the mempool became invalid and will be removed',
                               count=len(to_remove))
-                storage.remove_transactions(to_remove)
+                with storage.allow_invalid_context():
```
Comment (msbrogli) [inline hathor/consensus/consensus.py:116]: Add a note explaining that the mempool cleaning mechanism invalidates all transactions so we need to allow invalid transactions here.

File: hathor/transaction/storage/transaction_storage.py [hunk 12]
```diff
@@ -958,6 +1024,7 @@ def get_transactions_that_became_invalid(self) -> List[BaseTransaction]:
         for tx in self.iter_mempool_from_best_index():
             if tx.is_spent_reward_locked():
                 tx.get_metadata().validation = ValidationState.INVALID
```
Comment (msbrogli) [inline hathor/transaction/storage/transaction_storage.py:1026]: Should we refactor it to `tx.set_validation(ValidationState.INVALID)`? I feel we've been calling `mark_partially_validated()` in several places.

File: hathor/transaction/storage/tx_allow_scope.py (hunk 3 above)
Comment (msbrogli) [reply hathor/transaction/storage/tx_allow_scope.py:44]: Great! Thanks!

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/manager.py (hunk 4 above)
Comment (jansegre) [reply hathor/manager.py:445]: I think so.

File: hathor/manager.py (hunk 5 above)
Comment (jansegre) [reply hathor/manager.py:449]: That makes sense, I'm almost sure that it will be implicit as long as `get_all_transaction` respects the `tx_storage.allow_scope`.

File: hathor/manager.py (hunk 6 above)
Comment (jansegre) [reply hathor/manager.py:452]: Because `tx.update_initial_metadata()` does a `tx_storage.save_transaction(tx)` but we haven't validated `tx` yet. I don't like too much that some methods will implicitly save, but a lot of times it's hard to get around that. I think using `allow_partially_validated_context()` is correct here.

File: hathor/manager.py (hunk 7 above)
Comment (jansegre) [reply hathor/manager.py:486]: This is a bug fix. I'm not too sure why it was passing before, but `reject_locked_reward=False` should be correct because we aren't guaranteed to have a "good" sort when initialized (a good sort is a sort that validates the block that unlocks the reward before the transaction that spends the reward). But it is actually missing a final pass on the mempool to check for transactions that could have become invalid (those are the ones where `reject_locked_reward=True` must be used).

File: hathor/manager.py (hunk 8 above)
Comment (jansegre) [reply hathor/manager.py:493]: The structure was:

```
if can do a full validation of tx:
    do full validation of tx
else:
    do partial validation of tx
save tx
```

I refactored it to:

```
if can do a full validation of tx:
    do full validation of tx
    save tx
else:
    with allow partial context:
        do partial validation of tx
        save tx
```

File: hathor/manager.py (hunk 10 above)
Comment (jansegre) [reply hathor/manager.py:495]: Because it might need access to partially validated parents.

File: hathor/manager.py (hunk 9 above)
Comment (jansegre) [reply hathor/manager.py:497]: Because that transaction is only partially validated now, so I have to save inside an "allow partially validated" context.

File: hathor/transaction/storage/transaction_storage.py (hunk 12 above)
Comment (jansegre) [reply hathor/transaction/storage/transaction_storage.py:1026]: That would be good.

File: hathor/manager.py (hunk 8 above)
Comment (msbrogli) [reply hathor/manager.py:493]: Thanks for the explanation!

File: hathor/manager.py (hunk 9 above)
Comment (msbrogli) [reply hathor/manager.py:497]: Got it. Thanks!

File: hathor/consensus/consensus.py (hunk 11 above)
Comment (msbrogli) [reply hathor/consensus/consensus.py:120]: Great! Thanks!

File: hathor/manager.py (hunk 4 above)
Comment (msbrogli) [reply hathor/manager.py:445]: Great! Thanks!

File: hathor/manager.py (hunk 5 above)
Comment (msbrogli) [reply hathor/manager.py:449]: Great! Thanks!

File: hathor/manager.py (hunk 6 above)
Comment (msbrogli) [reply hathor/manager.py:452]: I agree. Thanks for the explanation.

File: hathor/transaction/storage/transaction_storage.py (hunk 12 above)
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py]: Great refactor. Thanks!

File: hathor/manager.py (hunk 10 above)
Comment (jansegre) [reply hathor/manager.py:491]: Since I fixed the cache clearing issue in the test it only needs the context to make the save now, and not the basic validation.

File: hathor/transaction/storage/transaction_storage.py [hunk 13]
```diff
@@ -563,10 +563,16 @@ def get_metadata(self, hash_bytes: bytes) -> Optional[TransactionMetadata]:
         except TransactionDoesNotExist:
             return None
 
+    def get_all_transactions(self) -> Iterator[BaseTransaction]:
+        """Return all transactions/blocks.
```
Comment (msbrogli) [inline hathor/transaction/storage/transaction_storage.py:567]: Suggested docstring: "Return all vertices (transactions and blocks) within the allowed scope."

File: hathor/manager.py (hunk 10 above)
Comment (msbrogli) [reply hathor/manager.py:491]: Great!

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/manager.py (hunk 7 above)
Comment (jansegre) [reply hathor/manager.py]: I'll do that, but ideally I think we should remove this code altogether possibly replacing it with two different approaches: 1) actually and explicitly rebuilding the database, 2) actually just verifying (no write operations) and only the last few blocks/transactions.

Comment (jansegre) [reply hathor/manager.py]: Done. Opened #627

**Review by msbrogli — APPROVED** [review]

File: hathor/manager.py (hunk 7 above)
Comment (msbrogli) [reply hathor/manager.py]: Great! Thanks!

File: hathor/transaction/storage/transaction_storage.py (hunk 13 above)
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py]: Thanks!

**Review by glevco — APPROVED** [review]
